File: 2019/day07/day07.py
#!/usr/bin/env python3
import sys
import itertools

if len(sys.argv) > 1:
    filename = sys.argv[1]
else:
    filename = "input.txt"

data = [int(i) for i in open(filename).read().split(",")]

class IntCodeVM:
    def __init__(self, memory):
        self.memory = memory[:]
        self.pc = 0
        self.inputs = []
        self.outputs = []
        self.halted = False

    def push_input(self, x):
        self.inputs.append(x)

    def run_until_output(self):
        while not self.halted and not self.outputs:
            self.step()

        if self.outputs:
            return self.outputs.pop(0)

    def step(self):
        if self.halted:
            return

        read = lambda i: self.memory[i] if len(self.memory) > i else 0

        op = read(self.pc)
        opcode = op % 100
        src_mode = (op % 1000) // 100
        trg_mode = (op % 10000) // 1000
        dst_mode = op // 10000

        src, trg, dst = read(self.pc + 1), read(self.pc + 2), read(self.pc + 3)
        op1 = src if src_mode == 1 else read(src)
        op2 = trg if trg_mode == 1 else read(trg)
        op3 = dst if dst_mode == 1 else read(dst)

        if opcode == 99:
            self.halted = True
            return
        elif opcode == 1:
            self.memory[dst] = op1 + op2
            self.pc += 4
        elif opcode == 2:
            self.memory[dst] = op1 * op2
            self.pc += 4
        elif opcode == 3:
            self.memory[src] = self.inputs.pop(0)
            self.pc += 2
        elif opcode == 4:
            self.outputs.append(op1)
            self.pc += 2
        elif opcode == 5:
            if op1 != 0:
                self.pc = op2
            else:
                self.pc += 3
        elif opcode == 6:
            if op1 == 0:
                self.pc = op2
            else:
                self.pc += 3
        elif opcode == 7:
            self.memory[dst] = int(op1 < op2)
            self.pc += 4
        elif opcode == 8:
            self.memory[dst] = int(op1 == op2)
            self.pc += 4
        else:
            raise Exception("Unknown opcode {}".format(op))

def run_amplifiers(program, settings):
    curr_input = 0
    for i in settings:
        vm = IntCodeVM(program)
        vm.push_input(i)
        vm.push_input(curr_input)
        curr_input = vm.run_until_output()
    return curr_input

# star1
max_signal = 0
for i in itertools.permutations(range(5)):
    signal = run_amplifiers(data, i)
    if signal > max_signal:
        max_signal = signal
print(max_signal)

def run_amplifiers_loop(program, settings):
    vms = []
    for i in settings:
        vm = IntCodeVM(program)
        vm.push_input(i)
        vms.append(vm)

    last_signal = 0
    curr_signal = 0
    while True:
        for i in range(len(vms)):
            vms[i].push_input(curr_signal)
            curr_signal = vms[i].run_until_output()
            if curr_signal == None:
                return last_signal
        last_signal = curr_signal
    return last_signal

# star 2
max_signal = 0
for i in itertools.permutations(range(5, 10)):
    signal = run_amplifiers_loop(data, i)
    if signal > max_signal:
        max_signal = signal
print(max_signal)

File: 2019/day07/day07.hs
#!/usr/bin/env stack
-- stack --resolver lts-14.16 script
{-# LANGUAGE ScopedTypeVariables #-}

import System.Environment (getArgs)
import Data.List.Split (splitOn)
import Control.Monad.ST.Lazy
import Data.Array
import Data.Array.ST
import Data.STRef.Lazy
import Data.List (maximum, permutations)

runProgram :: Array Int Int -> [Int] -> [Int]
runProgram arr i = runST $ do
  memory <- thaw arr
  inputs <- newSTRef i
  runProgram_ memory inputs

runProgram_
  :: forall s. STArray s Int Int
  -> STRef s [Int]
  -> ST s [Int]
runProgram_ memory inputs = exec 0
  where
    exec pc = do
      (opcode, (src, trg, dst), (op1, op2, op3)) <- fetchDecode pc
      case opcode of
        99 -> return []
        1  -> do
          writeMem dst (op1 + op2)
          exec (pc + 4)
        2  -> do
          writeMem dst (op1 * op2)
          exec (pc + 4)
        3  -> do
          val <- readInput
          writeMem src val
          exec (pc + 2)
        4  -> do
          (op1:) <$> exec (pc + 2)
        5  -> exec (if op1 /= 0 then op2 else pc + 3)
        6  -> exec (if op1 == 0 then op2 else pc + 3)
        7  -> do
          writeMem dst (fromEnum $ op1 < op2)
          exec (pc + 4)
        8  -> do
          writeMem dst (fromEnum $ op1 == op2)
          exec (pc + 4)
        _  -> fail $ "Unknown opcode " ++ show opcode
    readMem :: Int -> ST s Int
    readMem loc = do
      (begin, end) <- getBounds memory
      if begin <= loc && loc <= end then readArray memory loc else return 0
    writeMem :: Int -> Int -> ST s ()
    writeMem loc x = writeArray memory loc x
    fetchDecode pc = do
      op <- readMem pc
      let opcode = op `mod` 100
      let srcMode = (op `mod` 1000) `quot` 100
      let trgMode = (op `mod` 10000) `quot` 1000
      let dstMode = op `quot` 100000
      src <- readMem (pc + 1)
      trg <- readMem (pc + 2)
      dst <- readMem (pc + 3)
      op1 <- if srcMode == 0 then readMem src else return src
      op2 <- if trgMode == 0 then readMem trg else return trg
      op3 <- if dstMode == 0 then readMem dst else return dst
      return (opcode, (src, trg, dst), (op1, op2, op3))
    readInput = do
      i <- readSTRef inputs
      case i of
        []     -> return 0
        (x:xs) -> writeSTRef inputs xs >> return x

runAmplifiers :: Array Int Int -> [Int] -> Int
runAmplifiers prog settings = go settings [0]
  where go (s:ss) xs = go ss $ runProgram prog (s:xs)
        go []     xs = head xs

-- through the magic of haskell we do the loop by making a recursive binding!
-- lol wtf this actually just works with the lazy ST monad
runAmplifierLoop :: Array Int Int -> [Int] -> Int
runAmplifierLoop prog settings = let xs = go settings (0:xs) in last xs
  where go (s:ss) xs = go ss $ runProgram prog (s:xs)
        go []     xs = xs

star1 :: Array Int Int -> Int
star1 prog = maximum $ map (runAmplifiers prog) $ permutations [0..4]

star2 :: Array Int Int -> Int
star2 prog = maximum $ map (runAmplifierLoop prog) $ permutations [5..9]

main :: IO ()
main = do
  args <- getArgs
  let
    filename = case args of
      (x:_) -> x
      _     -> "input.txt"
  input <- map read . splitOn "," <$> readFile filename
  let inputArray = listArray (0, length input - 1) input
  print $ star1 inputArray
  print $ star2 inputArray
